ntp-daemon/src/system.rs

```diff
@@ -0,0 +1,199 @@
+use crate::peer::{MsgForSystem, PeerChannels, PeerIndex, PeerTask, ResetEpoch};
+use ntp_os_clock::UnixNtpClock;
+use ntp_proto::{
+    ClockController, ClockUpdateResult, FilterAndCombine, NtpInstant, PeerSnapshot, SystemConfig,
+    SystemSnapshot,
+};
+use tracing::info;
+
+use std::{error::Error, sync::Arc};
+use tokio::sync::{mpsc, watch};
+
+/// Spawn the NTP daemon
+pub async fn spawn(config: &SystemConfig, peer_addresses: &[&str]) -> Result<(), Box<dyn Error>> {
+    // shares the system state with all peers
+    let global_system_snapshot = Arc::new(tokio::sync::RwLock::new(SystemSnapshot::default()));
+
+    // send the reset signal to all peers
+    let reset_epoch: ResetEpoch = ResetEpoch::default();
+    let (reset_tx, reset_rx) = watch::channel::<ResetEpoch>(reset_epoch);
+
+    // receive peer snapshots from all peers
+    let (msg_for_system_tx, msg_for_system_rx) = mpsc::channel::<MsgForSystem>(32);
+
+    for (index, address) in peer_addresses.iter().enumerate() {
+        let channels = PeerChannels {
+            msg_for_system_sender: msg_for_system_tx.clone(),
+            system_snapshots: global_system_snapshot.clone(),
+            reset: reset_rx.clone(),
+        };
+
+        PeerTask::spawn(
+            PeerIndex { index },
+            address,
+            UnixNtpClock::new(),
+            *config,
+            channels,
+        )
+        .await?;
+    }
+
+    let mut peers = Peers::new(peer_addresses.len());
+
+    run(
+        config,
+        &mut peers,
+        reset_epoch,
+        global_system_snapshot,
+        msg_for_system_rx,
+        reset_tx,
+    )
+    .await
+}
+
+async fn run(
+    config: &SystemConfig,
+    peers: &mut Peers,
+    mut reset_epoch: ResetEpoch,
+    global_system_snapshot: Arc<tokio::sync::RwLock<SystemSnapshot>>,
+    mut msg_for_system_rx: mpsc::Receiver<MsgForSystem>,
+    reset_tx: watch::Sender<ResetEpoch>,
+) -> Result<(), Box<dyn Error>> {
+    let mut controller = ClockController::new(UnixNtpClock::new());
+    let mut snapshots = Vec::with_capacity(peers.len());
+
+    while let Some(msg_for_system) = msg_for_system_rx.recv().await {
+        peers.receive_update(msg_for_system, reset_epoch);
+
```

Comment:
I am missing the early out here for messages that should not cause a recalculation